doc/wg/core/notes/core-notes-2020-06-05.md

```diff
@@ -0,0 +1,96 @@
+ * Samuel: Probably. Although that assignment could be board-specific too.
+ * Leon: But does that make sense. Having a persistent ID for non-volatile board is useful for every board.
+ * Amit: A reasonable way for creating concise IDs at load time though, could just be creating a hash of the app.
+ * Andrey: But those app IDs should not be board specific, but application specific. The application decides what IDs they should take and use. Applications as in processes here. Some apps on the board might be loadable but others on the same board could be a static image with a well-known ID. A short, concise ID is the common denominator. How to get from a process image to that ID is board-specific.
+ * Leon: That makes sense. So a standard in-kernel format. Probably also a standard way its assigned and used for the standard boards that all do app loading the same way. Not every board needs to develop its own thing if it doesn't have a requirement.
+ * Amit: It sounds like a resounding consensus to not make the kernel do this, except for the application ID itself. But how to get them and verify them should be left up to the board, but with a strong understanding that for the moment most common use case will be "no verification". Because things might be done out-of-band with a secure loader, or dev boards that don't care. But we do want to support future use cases.
+ * Johnathan: Sounds good.
+ 
+ ## Unallow/Unsubscribe Loophole
+ * Amit: https://github.com/tock/tock/issues/1905
+ * Guillaume: Credit to Johnathan and Vadim who raised the issue. The subscribe/allow syscalls that give callbacks or slices and can later unallow or unsubscribe with a null pointer. The assumption in libtock-rs is that once you unsubscribe you have ownership again. I think this is going to be important for libtock-rs as its using lambda callback functions. And local buffers for slices. If we can't get back ownership, the lifetime is likely gone. The app might exit the function and the memory may not exist anymore. The kernel could end up corrupting the stack. So it's important that unallow exists. But the Tock kernel doesn't actually enforce that capsules do the right thing and give back ownership. I'll focus on slices but the problem is similar for both. We have to declare clear semantics about who has ownership of a slice. For example, a USB driver that wants to receive packets into userspace, usually you would allow a slice from userspace and subscribe to a callback then call a command to start it. Then the kernel would fill the slice and call the callback. After that userspace can interact with the data. It's important to define who owns the slice at what point in these operations. Because the kernel can't overwrite the slice with the next packet while the userland is interacting with it. So the solution in userland seems to be to allow and disallow slices in a double-buffer sense. But there is nothing enforcing that a capsule properly gives back the memory right now.
```

Comment:
Should be "Credit to Woyten and Johnathan who raised the original issue (https://github.com/tock/libtock-rs/issues/143)."

Reply:
```suggestion
 * Guillaume: Credit to Woyten and Johnathan who raised the issue (https://github.com/tock/libtock-rs/issues/143). The subscribe/allow syscalls that give callbacks or slices and can later unallow or unsubscribe with a null pointer. The assumption in libtock-rs is that once you unsubscribe you have ownership again. I think this is going to be important for libtock-rs as its using lambda callback functions. And local buffers for slices. If we can't get back ownership, the lifetime is likely gone. The app might exit the function and the memory may not exist anymore. The kernel could end up corrupting the stack. So it's important that unallow exists. But the Tock kernel doesn't actually enforce that capsules do the right thing and give back ownership. I'll focus on slices but the problem is similar for both. We have to declare clear semantics about who has ownership of a slice. For example, a USB driver that wants to receive packets into userspace, usually you would allow a slice from userspace and subscribe to a callback then call a command to start it. Then the kernel would fill the slice and call the callback. After that userspace can interact with the data. It's important to define who owns the slice at what point in these operations. Because the kernel can't overwrite the slice with the next packet while the userland is interacting with it. So the solution in userland seems to be to allow and disallow slices in a double-buffer sense. But there is nothing enforcing that a capsule properly gives back the memory right now.
```

Reply:
Thank you! I'm glad that didn't slip through.